src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshalling/AnsiStringMarshaller.cs

```diff
@@ -9,90 +9,109 @@ namespace System.Runtime.InteropServices.Marshalling
-                if (exactByteCount > buffer.Length)
+                _allocated = false;
+
+                if (managed is null)
                 {
-                    buffer = new Span<byte>((byte*)Marshal.AllocCoTaskMem(exactByteCount), exactByteCount);
-                    _allocated = true;
+                    _unmanagedValue = null;
+                    return;
                 }
-            }
 
-            _nativeValue = (byte*)Unsafe.AsPointer(ref MemoryMarshal.GetReference(buffer));
+                // >= for null terminator
+                // Use the cast to long to avoid the checked operation
+                if ((long)Marshal.SystemMaxDBCSCharSize * managed.Length >= buffer.Length)
+                {
+                    // Calculate accurate byte count when the provided stack-allocated buffer is not sufficient
+                    int exactByteCount = Marshal.GetAnsiStringByteCount(managed); // Includes null terminator
+                    if (exactByteCount > buffer.Length)
+                    {
+                        buffer = new Span<byte>((byte*)Marshal.AllocCoTaskMem(exactByteCount), exactByteCount);
+                        _allocated = true;
+                    }
+                }
 
-            Marshal.GetAnsiStringBytes(str, buffer); // Includes null terminator
-        }
+                _unmanagedValue = (byte*)Unsafe.AsPointer(ref MemoryMarshal.GetReference(buffer));
 
-        /// <summary>
-        /// Returns the native value representing the string.
-        /// </summary>
-        /// <seealso cref="CustomTypeMarshallerFeatures.TwoStageMarshalling"/>
-        public byte* ToNativeValue() => _nativeValue;
+                Marshal.GetAnsiStringBytes(managed, buffer); // Includes null terminator
+            }
 
-        /// <summary>
-        /// Sets the native value representing the string.
-        /// </summary>
-        /// <param name="value">The native value.</param>
-        /// <seealso cref="CustomTypeMarshallerFeatures.TwoStageMarshalling"/>
-        public void FromNativeValue(byte* value)
-        {
-            _nativeValue = value;
-            _allocated = true;
-        }
+            /// <summary>
+            /// Convert the current manage string to an unmanaged string.
+            /// </summary>
+            /// <returns>The unmanaged string</returns>
+            public byte* ToUnmanaged() => _unmanagedValue;
 
-        /// <summary>
-        /// Returns the managed string.
-        /// </summary>
-        /// <seealso cref="CustomTypeMarshallerDirection.Out"/>
-        public string? ToManaged() => Marshal.PtrToStringAnsi((IntPtr)_nativeValue);
+            /// <summary>
+            /// Initialize the marshaller with an unmanaged string.
+            /// </summary>
+            /// <param name="unmanaged">An unmanaged string</param>
+            public void FromUnmanaged(byte* unmanaged)
```

Comment:
We should not need `FromUnmanaged` and `ToManaged` method on the `In` mode marshaller.

Reply:
(same for other string marshallers)